src/ess/spectroscopy/indirect/io.py

```diff
@@ -0,0 +1,191 @@
+        fn = str(base) + '_' + f'{i+1}'.rjust(length, '0') + '.nxspe'
+        files.append(NXspeFileName(fn))
+        _to_one_nxspe(ev, fn, progress)
+    return NXspeFileNames(files)
+
+
+def _make_group(group: h5py.Group) -> scippnexus.Group:
+    return scippnexus.Group(group, definitions=scippnexus.base_definitions())
+
+
+def _lambda_to_ei(incident_wavelength):
+    from scipp.constants import Planck, neutron_mass
+
+    return ((Planck / incident_wavelength) ** 2 / neutron_mass / 2).to(unit='meV')
+
+
+def _ei_ef_to_en(incident_energy, final_energy):
+    return incident_energy - final_energy
+
+
+def _to_one_nxspe(events: DataArray, filename: str, progress):
+    """Use scippnexus to create the NXspe file"""
+    from scipp import full, scalar, sqrt
+    from scippnexus import (
+        NXcollection,
+        NXdata,
+        NXentry,
+        NXfermi_chopper,
+        NXinstrument,
+        NXsample,
+    )
+
+    observations = events.copy()
+    ef = events.coords['final_energy']
+    observations *= sqrt(events.bins.coords['incident_energy'] / ef)
+
+    # Adding null events requires replicating all coordinates of real events.
+    # We don't necessarily use all present event coordinates, so remove any we won't use
+    targets = [
+        'energy_transfer',
+        'incident_energy',
+        'incident_wavelength',
+        'final_energy',
+    ]
+    for coord in [x for x in events.bins.coords if x not in targets]:
+        del observations.bins.coords[coord]
+    # And provide a means to calculate the rest from bin (not event) information
+    # we are sure incident_wavelength is a bin coordinate already, so can skip it.
+    graph = {
+        'incident_energy': _lambda_to_ei,
+        'energy_transfer': _ei_ef_to_en,
+    }
+    # # Adding zero-weight observations, i.e., null events, works now, but involves
+    # # a memory-copy of the data array and its bin structure. Since it isn't strictly
+    # # necessary for exporting NXspe, it can be skipped for now.
+    # observations = add_null_observations(observations, targets, graph)
+
+    # combine the per bin intensities and normalize by monitor counts
+    # Note, applying this normalization to the _events_ would require splitting
+    # the bin monitor counts between the bin events.
+    normalize_by = events.coords['monitor']  # .broadcast(sizes=observations.sizes)
+    if normalize_by.variances is not None:
+        # we need to ignore the monitor uncertainty for the time being
+        normalize_by = normalize_by.copy()
+        normalize_by.variances = None
+
+    observations = observations.hist().transform_coords(targets, graph=graph)
+
+    if observations.variances is None:
+        observations.variances = observations.values  # correct for counting statistics
+        observations.variances[observations.values == 0] = 1
+    if observations.data.variances is not None:
+        observations.data.variances[observations.data.variances == 0] = 1
```

Comment:
What does this second `if` do? Isn't this covered by the first?

Reply:
Nothing :shrug: it's left-over from a refactor and I forgot to remove it. Good catch.